device.h: C++20 compatibility

marc-hb opened the pull request:
2 fixes + 2 corresponding `tests/` commits, please check commit messages.

Apply to device.h and pm/device.h the C/C++ compatibility lessons learned for init.h in #69411

marc-hb committed:
device.h: conditionally remove init braces for C++20 compatibility

Conditionally remove braces for designated initializers of anonymous
unions. This makes it compatible with C++20 while not breaking pre-C11
gcc.

This does for device.h what commit c15f029a7108 ("init.h: restore
designated initializers in SYS_INIT_NAMED()") did for init.h

See https://docs.zephyrproject.org/latest/develop/languages/cpp/ and
long discussion in #69411 for more obscure C/C++ compatibility details.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

marc-hb committed:
pm: device.h: fix pm_device designated initializers order for C++

Fixes the following error appearing in the test coverage added by the
next commit:

In file included from zephyr/tests/lib/cpp/cxx/src/main.cpp:18:
```
zephyr/include/zephyr/pm/device.h:275:9: error: designator order for
  field 'pm_device_base::state' does not match declaration order
                        in 'pm_device_base'
  275 |         }
      |         ^
zephyr/include/zephyr/pm/device.h:315:17: note: in expansion of
                                                macro 'Z_PM_DEVICE_INIT'
  315 |              Z_PM_DEVICE_INIT(Z_PM_DEVICE_NAME(dev_id), node_id,
```

Note this failure is observed with any g++ -std=c++NN standard value -
even before C++20.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

gmarull commented on include/zephyr/device.h:
```diff
@@ -906,9 +906,24 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
+ * support for anonymous unions but they require these braces when
+ * combined with C99 designated initializers. These braces are
+ * compatible with any C version but not with C++20.  For more obscure
+ * C/C++ compatibility details see commit c15f029a7108 and PR #69411.
```
please avoid referring to PRs in documentation

marc-hb replied:
I usually do but this issue is really complex and that PR discussion is quite important.

gmarull replied:
PRs content is not part of Zephyr git history. Who knows if Github will stay Zephyr's VCS provider forever.

fabiobaltieri replied:
I'm with Gerard on this, just refer to the commit, one can find back the PR out of band but really the commit should have enough context regardless

If you really want to link a PR maybe you can do it on this comment using the `Link:` tag and the full URL?

marc-hb replied:
> Who knows if Github will stay Zephyr's VCS provider forever.

I'm aware and I always try to avoid references to PRs when the git history is enough. But I think there can be some exceptions. It's not because something may disappear one day that it should not be referenced. If Zephyr migrates away from Github one day, then it should definitely export all past review discussions and migrate or archive them somewhere and then PR numbers would still be useful. I've been performing such archeology in the past and any reference helps.

Of course documentation is even better but (1) it's not mutually exclusive (2) not obvious where in the docs this obscure issue belongs (3) more "overhead" :-)

> PRs content is not part of Zephyr git history.

True but the git history does not exist in a complete vacuum either.

> If you really want to link a PR maybe you can do it on this comment using the Link: tag and the full URL?

That would work for me.

Or I can just delete the PR number entirely if it's really import for you. I think that would be a loss but it's not that big of a deal for me.

fabiobaltieri replied:
Full link of external references in the commit I think are fine, Linux does it as well, it's more my opinion than a general rule but I think it's a good idea and I occasionally do it myself too.

gmarull commented on include/zephyr/device.h:
```diff
@@ -980,12 +995,25 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
 	}
 
+/*
+ * Anonymous unions require C11. Some pre-C11 gcc versions have early
```
this is repeating information, can’t it be put somewhere and reference it?

marc-hb replied:
I tried to summarize it but I agree it's still a bit long and duplicated... any suggestion of a good place? I couldn't think of a really better way to document this sorry.

gmarull replied:
Maybe Zephyr coding guidelines? It looks like we'll need to apply this in other places if we want C++20 compatibility. An explanation there, plus a code snippet would be a solution, you could then just say "Refer to doc/xxx.rst"

marc-hb replied:
Thanks, I'll take a look in the next couple weeks.

marc-hb replied:
Done in #74047

marc-hb reviewed (commented):
Thx for the review! Answers below.

marc-hb committed:
tests: cpp: add C++ coverage for pm/device.h

Note CONFIG_PM_* affects device.h too. Even if not compiled, the more
code the pre-processor sees the better.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

marc-hb committed:
tests: cpp: add some C++ coverage for device.h / Device Tree

This provides C++ build test coverage for device.h (notably:
Z_DEVICE_INIT() fixed in the previous commit) and for some other Device
Tree macros.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

fabiobaltieri commented on include/zephyr/device.h:
```diff
@@ -906,7 +906,7 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
 		.data = (data_),                                               \
 		IF_ENABLED(CONFIG_DEVICE_DEPS, (.deps = (deps_),)) /**/        \
-		IF_ENABLED(CONFIG_PM_DEVICE, ({ .pm_base = (pm_),)}) /**/         \
+		IF_ENABLED(CONFIG_PM_DEVICE, ({ .pm_base = (pm_),})) /**/         \
```
I'm speachless

ceolin commented on include/zephyr/pm/device.h:
```diff
@@ -244,9 +244,9 @@ BUILD_ASSERT(offsetof(struct pm_device_isr, base) == 0);
-		.state = PM_DEVICE_STATE_ACTIVE,			     \
 		.flags = ATOMIC_INIT(Z_PM_DEVICE_FLAGS(node_id) | (_flags)), \
+		.state = PM_DEVICE_STATE_ACTIVE,			     \
+		.action_cb = pm_action_cb,				     \
```
is it needed ?

ceolin reviewed (commented):
+1 for adding the explanation in Zephyr's documentation. Please, don't forget to add the information about the need of initializers to be in order, this is probably affecting more places.

marc-hb commented:
> Maybe Zephyr coding guidelines? It looks like we'll need to apply this in other places if we want C++20 compatibility. An explanation there, plus a code snippet would be a solution, you could then just say "Refer to doc/xxx.rst"

> +1 for adding the explanation in Zephyr's documentation. 

Submitted:
- #74047

marc-hb commented:
I believe all review comments should have been addressed, please review again.

There's even more coverage now. It's been time-consuming mainly due to my Device Tree ignorance (partially fixed) but at last this PR adds a significant amount of C++ coverage for a relatively small number of lines. I think it's good value.

fabiobaltieri commented on include/zephyr/device.h:
```diff
@@ -1080,11 +1080,22 @@ device_get_dt_nodelabels(const struct device *dev)
+ * more details see https://docs.zephyrproject.org/latest/develop/languages/cpp/
+ */
+#if defined(__STDC_VERSION__) && (__STDC_VERSION__) < 201100
+#  define Z_DEVICE_INIT_PM_BASE(pm_) ({ .pm_base = (pm_),})
```
most nested define in the code base are not indented, I would keep this aligned for consistency, up to you I guess

marc-hb replied:
True, such indentation is already done in very few places but I think it's mostly because people just don't know they can. Or they are not sure and err on the overly conservative side - even though there has never been any issue ever reported related to this. I haven't seen any Zephyr style rule about this. I'm all for consistency when there are small variations to do the same thing with none clearly better than the other one. But in this case the two alternatives are not equivalent, they are: (1) add useful information, versus (2) not add useful information.

It makes little difference here but it's _very_ helpful as soon as you have more than 2-3 levels of nesting. Just imagine not indenting actual C code. So I believe it should become more popular.

marc-hb commented:
Tentative v3.7.0 milestone because this PR:
- **fixing** C++ compatibility
- adding **tests**.

Absolutely **zero feature** change.

nashif commented:
@gmarull  please review

keith-packard commented on include/zephyr/device.h:
```diff
@@ -1169,12 +1177,23 @@ device_get_dt_nodelabels(const struct device *dev)
+#  define Z_DEVICE_INIT_ENTRY_DEV(node_id, dev_id)   Z_DEV_ENTRY_DEV(node_id, dev_id)
+#endif
+
+#define Z_DEV_ENTRY_DEV(node_id, dev_id)                                                           \
```
Odd to find this macro defined after it was "used" above. Not strictly required, of course.

marc-hb replied:
This places the macro closer to the two locations where it's used.

It's also a lower-level implementation detail, so I think it it reads better last - that's subjective of course.

I think it's "odd" only if you're tainted by C compilation which requires "forward" declarations. Maybe dating all the way back to when compilation was performed in a single pass? Funny how the pre-processor is in some respect more advanced that C itself... https://stackoverflow.com/questions/3136686/is-the-c99-preprocessor-turing-complete

gmarull commented on include/zephyr/device.h:
```diff
@@ -1080,11 +1080,22 @@ device_get_dt_nodelabels(const struct device *dev)
+ * more details see https://docs.zephyrproject.org/latest/develop/languages/cpp/
+ */
+#if defined(__STDC_VERSION__) && (__STDC_VERSION__) < 201100
+#  define Z_DEVICE_INIT_PM_BASE(pm_) ({ .pm_base = (pm_),})
```
isn't this missing a trailing `,`?

gmarull replied:
see https://github.com/zephyrproject-rtos/zephyr/pull/74970

marc-hb replied:
Great catch, thank you @gmarull for the fix in #74970. Note this regression is only 3 weeks old and independent of this PR. My bad though for not noticing it when I rebased on top of it.

So I just:
- rebased on top of your comma fix #74970 
- Added `CONFIG_DEVICE_DT_METADATA` to the `prj.conf`, on top of the other additions that were already in this PR. I checked it would have caught this bug.

marc-hb reviewed (commented):
Just force-pushed:
- Rebased on top of comma fix #74970 fixing 3 weeks old regression from #71912
- Added `CONFIG_DEVICE_DT_METADATA` to the `prj.conf`, on top of the other CONFIG_ additions that were already in this PR. It would have caught this "foreign" regression.

NO OTHER CHANGE.

Please re-review this very minor update quickly before I have to rebase on top of another regression and then another fix :-)

Sorry for the inconvenience.

marc-hb commented:
> [minor changes]
> NO OTHER CHANGE.
> Please re-review this very minor update quickly before I have to rebase on top of another regression and then another fix :-)
> Sorry for the inconvenience.

Could I get just one of the previous approvers to take another, quick look? Thanks in advance!
